File: Javascript/tic-tac-toe.js
var prompt = require('prompt-sync')();

let board = [0]
for (let i = 0; i < 9; i++) {
	board[i] = 0;
}

let player1Turn = true;

while (!checkGameWin()) {
	if (checkDraw(board)) {
		printBoard(board)
		console.log("This game was draw")
		return 0;
	}

	console.log("Please take a turn player " + (player1Turn ? "1" : "2"))
	console.log("The board looks like this:")
	printBoard(board);

	let input = prompt("Type the position where you want to place your mark! [1-9] ");
	while (Number.isNaN(Number(input)) || input < 1 || input > 9) {
		// check for CTRL-C
		if (input === null) {
			process.exit(0);
		}

		console.log("Incorrect input. Please only type characters from 1 to 9");
		input = prompt("Type the position where you want to place your mark! [1-9] ");
	}

	let index = input - 1;

	// if the field isn't already occupied
	if (board[index] === 0) {
		// set field to correct state depending on who's turn it is
		board[index] = player1Turn ? 1 : 2;

		// other player's turn
		player1Turn = !player1Turn;
	}
}

printBoard(board);
console.log("Player " + (!player1Turn ? "1" : "2") + " won the game!")

function checkGameWin() {
	// check horizontal
	for (let i = 0; i < 3; i++) {
		// check X
		if (board[0 + i * 3] === 1 && board[1 + i * 3] === 1 && board[2 + i * 3] === 1) {
			return true;
		}

		// check O
		if (board[0 + i * 3] === 2 && board[1 + i * 3] === 2 && board[2 + i * 3] === 2) {
			return true;
		}
	}

	// check vertical
	for (let i = 0; i < 3; i++) {
		// check X
		if (board[i] === 1 && board[i + 3] === 1 && board[i + 6] === 1) {
			return true;
		}

		// check O
		if (board[i] === 2 && board[i + 3] === 2 && board[i + 6] === 2) {
			return true;
		}
	}

	// check diagonal X
	if ((board[0] === 1 && board[4] === 1 && board[8] === 1) 
		|| board[2] === 1 && board[4] === 1 && board[6] === 1) {
		return true;
	}

	// check diagonal O
	if ((board[0] === 2 && board[4] === 2 && board[8] === 2) 
		|| board[2] === 2 && board[4] === 2 && board[6] === 2) {
		return true;
	}

	return false;
}

function checkDraw() {
	for (const field of board) {
		if (field === 0) {
			return false;
		}
	}

	return true;
}

function printBoard() {
	for (let i = 0; i < 3; i++) {
		console.log("+-+-+-+");
		let line = ""
		for (let j = 0; j < 3; j++) {
			line += '|';
			if (board[i * 3 + j] === 0) {
				line += ' ';
			}
			else if (board[i * 3 + j] === 1) {
				line += 'X';
			}
			else if (board[i * 3 + j] === 2) {
				line += 'O';
			}
		}
		console.log(line + '|')
	}
	console.log("+-+-+-+");
}

File: Python/tic-tac-toe.py
board = [[0,0,0],[0,0,0],[0,0,0]]
player_1_turn:bool = True

def checkGameWin(b, state: int) -> bool:
	for i in range(0, 3):
		if b[i][0] == state and b[i][1] == state and b[i][2] == state:
			return True
		
		if b[0][i] == state and b[1][i] == state and b[2][i] == state:
			return True
		
	# diagonal1
	if b[0][0] == state and b[1][1] == state and b[2][2] == state:
		return True

	# diagonal2
	if b[2][0] == state and b[1][1] == state and b[0][2] == state:
		return True
		
	return False

def checkDraw(board):
	for a in board:
		if 0 in a:
			return False
	return True

def printBoard(board):
	def b(i,j):
		if board[i][j] == 1:
			return "X"
		elif board[i][j] == -1:
			return "O"
		else:
			return " "
		
	for i in range(0, 3):
		print("+-+-+-+")
		print(f"|{b(i,0)}|{b(i,1)}|{b(i,2)}|")
	print("+-+-+-+")

while (not checkGameWin(board, -1) and not checkGameWin(board, 1)):
	if checkDraw(board):
		printBoard(board)
		print("The game was a draw!")
		break

	print(f"Please take a turn player {1 * (not player_1_turn) + 1}")
	print("The board looks like this:")
	printBoard(board)
	
	index: int = 0
	while True:
		inp = input("Type the position where you want to place your mark! [1-9]")
		if inp.isdigit() and int(inp) in range(1, 10):
			index = int(inp) - 1
			break
		print("Incorrect input. Please only type characters from 1 to 9")
	
	if board[index//3][index%3] == 0:
		board[index//3][index%3] = 1 if player_1_turn else -1
		player_1_turn = not player_1_turn

printBoard(board)
print(f"Player {1 * player_1_turn + 1} won this game!")